proxy.py

```diff
@@ -1093,24 +1107,38 @@ def name(self) -> str:
         return self.__class__.__name__      # pragma: no cover
 
     @abstractmethod
-    def before_upstream_connection(self) -> bool:
+    def before_upstream_connection(self, request: HttpParser) -> Optional[HttpParser]:
```

Comment:
@knucker @moham96 Updated plugin definition, added some comments to clarify when these callbacks are invoked and how they can drive the behavior.